src/hierarchies/base_hierarchy.h

```diff
@@ -29,101 +29,312 @@
+    const Eigen::MatrixXd &data,
+    const Eigen::MatrixXd &covariates /*= Eigen::MatrixXd(0, 0)*/) {
+  Eigen::VectorXd lpdf(data.rows());
+  if (covariates.cols() == 0) {
+    for (int i = 0; i < data.rows(); i++) {
+      // Pass null value as covariate
+      lpdf(i) = static_cast<Derived *>(this)->like_lpdf(data.row(i),
+                                                        Eigen::RowVectorXd(0));
+    }
+  } else {
+    for (int i = 0; i < data.rows(); i++) {
+      lpdf(i) = static_cast<Derived *>(this)->like_lpdf(data.row(i),
+                                                        covariates.row(i));
+    }
+  }
+  return lpdf;
+}
+
+template <class Derived, typename State, typename Hyperparams, typename Prior>
+Eigen::VectorXd
+BaseHierarchy<Derived, State, Hyperparams, Prior>::prior_pred_lpdf_grid(
+    const Eigen::MatrixXd &data,
+    const Eigen::MatrixXd &covariates /*= Eigen::MatrixXd(0, 0)*/) {
+  Eigen::VectorXd lpdf(data.rows());
+  if (covariates.cols() == 0) {
+    for (int i = 0; i < data.rows(); i++) {
+      // Pass null value as covariate
+      lpdf(i) = static_cast<Derived *>(this)->prior_pred_lpdf(
+          data.row(i), Eigen::RowVectorXd(0));
+    }
+  } else {
+    for (int i = 0; i < data.rows(); i++) {
+      lpdf(i) = static_cast<Derived *>(this)->prior_pred_lpdf(
+          data.row(i), covariates.row(i));
+    }
+  }
+  return lpdf;
+}
+
+template <class Derived, typename State, typename Hyperparams, typename Prior>
+Eigen::VectorXd
+BaseHierarchy<Derived, State, Hyperparams, Prior>::conditional_pred_lpdf_grid(
+    const Eigen::MatrixXd &data,
+    const Eigen::MatrixXd &covariates /*= Eigen::MatrixXd(0, 0)*/) {
+  Eigen::VectorXd lpdf(data.rows());
+  if (covariates.cols() == 0) {
+    for (int i = 0; i < data.rows(); i++) {
+      // Pass null value as covariate
+      lpdf(i) = static_cast<Derived *>(this)->conditional_pred_lpdf(
+          data.row(i), Eigen::RowVectorXd(0));
+    }
+  } else {
+    for (int i = 0; i < data.rows(); i++) {
+      lpdf(i) = static_cast<Derived *>(this)->conditional_pred_lpdf(
+          data.row(i), covariates.row(i));
+    }
+  }
+  return lpdf;
+}
+
+template <class Derived, typename State, typename Hyperparams, typename Prior>
+void BaseHierarchy<Derived, State, Hyperparams, Prior>::sample_full_cond(
+    const Eigen::MatrixXd &data,
+    const Eigen::MatrixXd &covariates /*= Eigen::MatrixXd(0, 0)*/) {
+  static_cast<Derived *>(this)->clear_data();
+  if (covariates == Eigen::MatrixXd(0, 0)) {
+    for (int i = 0; i < data.rows(); i++)
+      static_cast<Derived *>(this)->add_datum(i, data.row(i), false,
+                                              Eigen::RowVectorXd(0));
+  } else {
+    for (int i = 0; i < data.rows(); i++)
+      static_cast<Derived *>(this)->add_datum(i, data.row(i), false,
+                                              covariates.row(i));
+  }
+  static_cast<Derived *>(this)->sample_full_cond();
```

Comment:
This static_cast construct occurs quite a bit and is very ugly and verbose, is there any possibility of shortening it? maybe with a `using` declaration or something?

Reply:
I'm not sure since it requires `this` to be available

Reply:
Ok, we'll leave it as is. It's not like it matters much.